libraries/extractor/src/main/java/androidx/media3/extractor/ts/AdtsReader.java

```diff
@@ -192,10 +191,11 @@ public void consume(ParsableByteArray data) throws ParserException {
           }
           break;
         case STATE_READING_AAC_PCE:
-          checkNotNull(pceBuffer);
-          if (continueRead(data, pceBuffer.data, pceBuffer.data.length)) {
-            readAacProgramConfigElement();
-          }
+          int offset = data.getPosition();
+          int limit = offset + min(sampleSize, AAC_PCE_MAX_SIZE);
+          ParsableBitArray pceBuffer =
+              new ParsableBitArray(Arrays.copyOfRange(data.getData(), offset, limit));
```

Comment:
How does this work when `data` doesn't contain the entire PCE? If you don't use `continueRead` to buffer the PCE up then you need some sort of byte-by-byte state tracking inside `readAacProgramConfigElement`, which you don't have.

Reply:
What do you mean by entire PCE?

Instead of calling `continueRead` which moved the `data` position further ahead to read `pceBuffer`, we are now copying the `pceBuffer` from the `data` without changing its position. Once PCE data is read and appended to the `initializationData` of  `format`, `setReadingState` will be triggered which reads the sample and moves the position in data as we would have done when this method was not in place. 

We're now copying the data only once here instead of doing it twice in the previous approach. Please explain if you have something else in mind with this comment.

Reply:
Also instead of relying on `IllegalStateException` when `format` is not set for the case when `channelConfig` is `0` and not enough bytes in `PCE`, we now throw `ParserException` as PCE data should be present in this case.

Reply:
I understand why you've done that as it does simplify the code. It just doesn't work.

Take this test for example:
```
  @Test
  public void aacPceSplit() throws ParserException {
    byte[] first = Arrays.copyOf(AAC_PCE_TEST_DATA, AAC_PCE_ADTS_HEADER.length + 1);
    byte[] second = Arrays.copyOfRange(AAC_PCE_TEST_DATA,AAC_PCE_ADTS_HEADER.length + 1, AAC_PCE_TEST_DATA.length);

    data = new ParsableByteArray(first);
    feed();
    data = new ParsableByteArray(second);
    feed();

    assertSampleCounts(0, 1);
    adtsOutput.assertSample(0, AAC_PCE_ADTS_CONTENT, 0, C.BUFFER_FLAG_KEY_FRAME, null);
  }
```
This test passes enough data to cover just the ADTS header and the first byte of the PCE, then follows it up with the rest of the PCE. This doesn't work with your change.

Without storing the entire PCE in a buffer (using `continueRead`), you've lost the ability to break up the PCE into multiple calls to `consume`.

Imagine getting rid of `continueRead` from the `STATE_READING_ADTS_HEADER` case. It would break every time an ADTS header staddles the boundary of multiple `consume` calls. It's the same issue here but with the PCE.

Reply:
Thank you for the explanation. Makes sense to me now. I'll also add this test case now.

Do you agree with the change that once `readAacProgramConfigElement` is called and not enough bits are present or the PCE tag is not present, it is better to throw `ParserException` instead of silently not setting the format and relying on later failing because of this?

Reply:
> Do you agree with the change that once `readAacProgramConfigElement` is called and not enough bits are present or the PCE tag is not present, it is better to throw `ParserException` instead of silently not setting the format and relying on failing later because of this?

This is similar to how I first implemented this. If the PCE reading fails first time, I was just submitting the original 2 byte format and carrying on as normal. This was unreliable when seeking the stream because not all samples seem to have the PCE in. I then changed the code to hold the format back until a PCE is discovered. Therefore I expect this change to throw an exception periodically.

Reply:
I didn't quite understand if the new way of throwing error instead of silently failing is correct or not from the reply.

I've updated my commit, can you please point out if you still see some logical errors in parsing or let me know if this looks good to you.

Thanks

Reply:
The problem is that even though a PCE has to be at the start of a sample (it cannot be later in a sample), not every sample has a PCE. You might get the PCE just once per TS file, for example. Or perhaps once every N samples. This is a problem when seeking or otherwise consuming data from the middle of a stream.

Right now you're throwing a `ParserException` in three places in `readAacProgramConfigElement`. I think the last two are fine because they should only be reached if it really is a malformed stream. But the first exception you're throwing can be legitimately reached, when the stream is consumed starting from a sample that is missing a PCE. In those situations, my previous code kept trying every sample until a PCE is found.

I think you need to remove that first exception and just pass on the sample data as normal. Then when the next sample starts, it'll try to parse a PCE again.

Reply:
Makes sense. I guess when passing the sample data as normal, we also need to clear `pendingOutputFormat` to be on the safe side along with the `pceBuffer` as new `pendingOutputFormat` will be read from the header of the next sample. 

Thanks for the input. I have modified the changes accordingly.

Reply:
@lawadr,

> You might get the PCE just once per TS file, for example. Or perhaps once every N samples. This is a problem when seeking or otherwise consuming data from the middle of a stream.

The spec for this says : ` If channel_configuration equals 0, the channel configuration is not specified in the header and must be given by a program_config_element() following as first syntactic element in the first raw_data_block() after the header`.

So if we are inside `readAacProgramConfigElement`, we have read the header just before it and channel configuration is 0. Without the format how can we even consume the sample without a `format` information?

Also based on the spec, the channel configuration is part of the fixed adts header and will be same in all the headers, so if one `adts` header says `channel configuration` is `0`, the rest of the headers would do the same and it should be followed by PCE data in the raw blocks.  

> But the first exception you're throwing can be legitimately reached, when the stream is consumed starting from a sample that is missing a PCE.

Can you please shed more light on how this can be legitimately reached? I think it is safe to throw an exception even in this case. Also I couldn't find anything about PCE just present once per file, I think in that case if we start from the middle we're gonna crash anyways right?

Reply:
@rohitjoins 
> Also based on the spec, the channel configuration is part of the fixed adts header and will be same in all the headers, so if one adts header says channel configuration is 0, the rest of the headers would do the same and it should be followed by PCE data in the raw blocks.

This isn't true. Once a PCE is given inside a `raw_data_block`, it applies from that point onwards until another PCE is defined inside another `raw_data_block`. It says so in the specification:

_"The configuration indicated by the PCE takes effect at the raw_data_block() containing the
PCE. The number of front, side and back channels as specified in the PCE must be present
in that block and all subsequent raw_data_block()'s until a raw_data_block() containing a
new PCE is transmitted."_

Typically, as streams do not tend to change their channel setups, you'll only see one PCE at the very start. This is what I see in my sample. The first block starts with a PCE, while the following blocks jump straight into the SCEs and CPEs.

> Can you please shed more light on how this can be legitimately reached? I think it is safe to throw an exception even in this case. Also I couldn't find anything about PCE just present once per file, I think in that case if we start from the middle we're gonna crash anyways right?

So this brings us to TS files, specifically as part of an HLS playlist. If a PCE is present at the start of the first AAC block in a particular TS file, that TS file can be played in isolation and used as the starting point of the HLS.

You're right that if a PCE has not been encountered then samples cannot be decoded and we should throw. However, the issue is that we are determining whether to enter `readAacProgramConfigElement` based on the `hasOutputFormat` member of the `AdtsReader` class. The problem with that is if you play an HLS from the start and skip far enough ahead to another TS file, a new `AdtsReader` will be constructed and `hasOutputFormat` will be initialised to `false` again, even though we'll be reusing the previously created `TrackOutput` that has already received the format from before. In other words, I think we need to base whether to go into `readAacProgramConfigElement` on whether the `currentOutput` has previously received a format.

Reply:
Thank you for the explanation. Can you please add a commit which makes the changes you suggested above?

Reply:
I guess in that case we will be entering `readAacProgramConfigElement` only when we expect PCE data to be present and can throw an error when non-PCE tag (!= 5) is encountered.

Reply:
> Taking a closer look and discussing it internally, it seems you can't really check if currentOutput has a set Format or not.

Yes. This is a problem.

> The quoted documentation only makes sense within a single adts frame. **From what we understand, it seems every adts frame should have a header and n number of raw blocks.** Ts container is made for broadcasting and it is supposed to resume playback from the start of any frame in the sample. If the PCE data is only present at the start of first frame, playback would not be possible if you're starting playback from a frame which is not the start.

It does have a header and a number of raw blocks. You can see that in `adts_frame` in the spec. There's also something defined above `adts_frame` though, and that's `adts_sequence` which loosely defines a series of frames. I can see nothing that indicates a PCE only applies to the end of its `adts_frame` instead of `adts_sequence`.

> To reach a conclusion and merge this PR, I would request you to share the HLS stream (if possible) so that we can debug the sample on our end and understand what exactly is happening. The current file sent to us is just a sample file and I cannot reproduce the scenarios as stated above i.e. seeking inside a HLS stream.

I've got a test HLS here that exhibits this behaviour. It's 1.2gb though so I'm not sure how to send it across.

If you have ffmpeg you can create it yourself in no time:
1. Download an MP4 of Blender's Sprite Fright short: https://dl.acm.org/doi/10.1145/3550339.3556000
2. `ffmpeg -i <input.mp4> -c:v copy -c:a aac -aac_pce 1 -f hls -hls_list_size 0 <output.m3u8>`

This HLS plays fine in VLC, for example. You can play it from the beginning, skip forward and backwards. You can even start playback from a position that has no PCE using the `--start-time=` command line argument. I guess it either always checks the first TS file in the playlist to find the PCE or it works out the channel layout implicitly (see _8.5.2.3 Implicit channel mapping_ in the spec)?

If we throw whenever we can't find a PCE then you can only play the stream from the start and can _never_ skip forward or backwards. If we don't throw (or somehow check whether `currentOutput` already has a format) then you _can_ skip forward and backwards as long as the stream was played from the start. You still won't be able to start part way through the stream unless it's reencoded to replicate the PCE in every frame, we parse the stream from the start until the PCE is found, or we implicitly work out the PCE as described in the spec.